feature: Initialize kn admin plugin

zhanggbj opened the pull request:
Hi All,

This kn-admin plugin is designed to help administrators or operators better manage a Knative platform installation with kn CLI. 

This is an initial commit. More details in the proposal: https://docs.google.com/document/d/1oD5P4WQZs39GIr_4ezcURgRCuUx8dXrcCi3N7GKmp5c/edit#

Thanks!

zhanggbj commented:
@rhuss 
Don't worry, take your time:-)

rhuss commented:
@zhanggbj sorry, I couldn't make it before my PTO (and I'm heading now off to the Alpes with little connectivity). I will be back the first week in March and get your PRs merged then is on the top of my priority list.

Apologies, and have a nice week !

maximilien commented:
/ok-to-test

maximilien commented:
Hi, @zhanggbj, so all the tests are failing :(

I’ll leave some comments as well. But please fix this ^^^ first. Thanks.

maximilien commented on plugins/kn-admin/README.md:
```diff
@@ -0,0 +1,66 @@
+# kn-admin
+A plugin for kn client for management
```
Change to “A kn plugin for Knative cluster management”

zhanggbj replied:
Fixed!

maximilien commented on plugins/kn-admin/README.md:
```diff
@@ -0,0 +1,66 @@
+# kn-admin
+A plugin for kn client for management
+This kn-admin plugin is designed to help administrators or operators better manage a Knative platform installation with kn CLI. 
```
Change “or” to “and”

maximilien commented on plugins/kn-admin/README.md:
```diff
@@ -0,0 +1,66 @@
+A plugin for kn client for management
+This kn-admin plugin is designed to help administrators or operators better manage a Knative platform installation with kn CLI. 
+The plugin’s main objective is to make administration and operation workflows easier, for instance by making it easy to accomplish 
+tasks such as feature flags enablement or disablement with one command,instead of many manual steps like modifying ConfigMaps or yaml files.
```
s/command,instead/command, instead/

maximilien commented on plugins/kn-admin/README.md:
```diff
@@ -0,0 +1,66 @@
+Available Commands:
+  domain           Manage route domain
+  help             Help about any command
+  private-registry Manage private-registry
```
Might be good to add `version` command

zhanggbj replied:
I'll add it in next PR and follow kn client version cmd.

maximilien commented on plugins/kn-admin/README.md:
```diff
@@ -0,0 +1,66 @@
+  private-registry Manage private-registry
+
+Flags:
+      --config string   config file (default is $HOME/.kn-admin.yaml)
```
See the new XDG_... new fashion for config that is being added to `kn`

zhanggbj replied:
Fixed!
--config string   config file (default is $HOME/.config/kn/plugins/kn-admin.yaml)

maximilien commented on plugins/kn-admin/cmd/domain/domain.go:
```diff
@@ -0,0 +1,39 @@
+	"github.com/spf13/cobra"
+)
+
+// domainCmd represents the domain command
```
I think you need to repeat the name of the function, e.g., “NewDomainCmd” as the first word in comment

maximilien commented on plugins/kn-admin/cmd/domain/domain.go:
```diff
@@ -0,0 +1,39 @@
+		Long: `List and set default route domain or route domain for Service with selectors. For example:
+
+kn admin domain list - to list Knative route domain
+kn admin domain set - to set Knative route domain`,
```
But with `set` would you need to also pass an argument?

maximilien commented on plugins/kn-admin/cmd/domain/set.go:
```diff
@@ -0,0 +1,107 @@
+var domain string
+
+// updateCmd represents the update command
+
```
No new line and same comment as above for doc comments

maximilien commented on plugins/kn-admin/cmd/domain/set.go:
```diff
@@ -0,0 +1,107 @@
+				os.Exit(1)
+			}
+
+			desiredCm := currentCm.DeepCopy()
```
Can we split this function into smaller private functions with clear name to what is going on :) thanks

maximilien replied:
Similar to the `splitByEqualSign`. There are three immediate that come to mind.

maximilien commented on plugins/kn-admin/cmd/private-registry/enable.go:
```diff
@@ -0,0 +1,125 @@
+	Email    string `json:"Email"`
+}
+type Auths struct {
+	UsIcrIo UsIcrIo `json:"us.icr.io"`
```
Hmm, is this fix `us.icr.io`? Could it change?

maximilien commented on plugins/kn-admin/cmd/private-registry/enable.go:
```diff
@@ -0,0 +1,125 @@
+and usage of using your command. For example:
+
+Cobra is a CLI library for Go that empowers applications.
+This application is a tool to generate the needed files
```
I don’t see the value of these three lines starting with “Cobra ...”

maximilien commented on plugins/kn-admin/cmd/private-registry/enable.go:
```diff
@@ -0,0 +1,125 @@
+Cobra is a CLI library for Go that empowers applications.
+This application is a tool to generate the needed files
+to quickly create a Cobra application.`,
+		Run: func(cmd *cobra.Command, args []string) {
```
Same as above. Split into smaller private funcs to this is readable at a glance.

maximilien commented on plugins/kn-admin/cmd/private-registry/enable.go:
```diff
@@ -0,0 +1,125 @@
+
+			j, err := json.Marshal(dockerCfg)
+			if err != nil {
+				panic(err)
```
We should consistently handle errors that result in panic by calling a common function. Also ensure to give user some message. I would also rather os.Exit(1) rather than just panic.

maximilien commented on plugins/kn-admin/cmd/private-registry/private-registry.go:
```diff
@@ -0,0 +1,39 @@
+		Long: `A longer description that spans multiple lines and likely contains examples
+and usage of using your command. For example:
+
+Cobra is a CLI library for Go that empowers applications.
```
Same as above. These three lines are not needed.

zhanggbj replied:
Fixed!

maximilien commented on plugins/kn-admin/cmd/private-registry/private-registry.go:
```diff
@@ -0,0 +1,39 @@
+Cobra is a CLI library for Go that empowers applications.
+This application is a tool to generate the needed files
+to quickly create a Cobra application.`,
+		Run: func(cmd *cobra.Command, args []string) {
```
Why is there an empty `Run`?

zhanggbj replied:
Fixed!

maximilien commented on plugins/kn-admin/main.go:
```diff
@@ -0,0 +1,22 @@
+import "github.com/knative/client-contrib/plugins/kn-admin/core"
+
+func main() {
+  core.NewAdminCommand().Execute()
```
What if `Execute()` fails? You need to handle error?

zhanggbj replied:
Good catch! Fixed!

maximilien commented on plugins/kn-admin/hack/build.sh:
```diff
@@ -0,0 +1,41 @@
+#!/bin/bash
+set -o pipefail
```
Is there a way to handle running tests?

zhanggbj replied:
Yes, as we have no UT so far and will add test later.

maximilien commented on plugins/kn-admin/core/root.go:
```diff
@@ -0,0 +1,99 @@
+`,
+	}
+	cobra.OnInitialize(initConfig)
+	rootCmd.PersistentFlags().StringVar(&cfgFile, "config", "", "config file (default is $HOME/.kn-admin.yaml)")
```
Note command above with about new config fashion

maximilien replied:
Also pay attention to how to deal with missing config file and how defaults are added. See kn version

zhanggbj replied:
Updated with new config fashion.

zhanggbj commented:
@maximilien Thank you for you review!

I've address your most of your comments and left one comment (add version in next PR), would you please help to take a review, thx!

I check the test failure, looks like it doesn't relate do plugin code, it cannot find the proper test file to execute. For e.g.

```
+ ./test/presubmit-tests.sh --integration-tests
./test/presubmit-tests.sh: line 37: ./test/../vendor/knative.dev/test-infra/scripts/presubmit-tests.sh: No such file or directory
./test/presubmit-tests.sh: line 40: main: command not found
+ EXIT_VALUE=127
+ set +o xtrace
```

maximilien commented:
Hi, @zhanggbj I think you need to add the test files in your PR. Doing that too for my PR #4.

I guess which ever gets merged first will have these files set and we can resolve conflicts in subsequent PRs.

Please do that and get all green and I will do a final pass. Thanks.

rhuss commented on go.mod:
```diff
@@ -0,0 +1,14 @@
+module github.com/knative/client-contrib
```
there should be no top-level go.mod but each individual plugin uses its own dependency tree. the same is true for the `vendor/` dir. See the comments to #5 for the reasoning and how I suggest to progress.

zhanggbj replied:
@rhuss sure, I'm working on it, will move vendor to the admin own structure and resubmit.

rhuss commented:
/retest

mattmoor commented on plugins/admin/cmd/kn-admin.go:
```diff
@@ -0,0 +1,27 @@
+package main
+
+import (
+	"fmt"
```
<!--gotool-bot-->
Format Go code:
```suggestion
	"fmt"
	"os"

```

mattmoor commented on plugins/admin/cmd/kn-admin.go:
```diff
@@ -0,0 +1,27 @@
+import (
+	"fmt"
+	"knative.dev/client-contrib/plugins/admin/core"
+	"os"
```
<!--gotool-bot-->
Format Go code:
```suggestion

```

mattmoor commented on plugins/admin/core/root.go:
```diff
@@ -0,0 +1,99 @@
+package core
+
+import (
+	"fmt"
```
<!--gotool-bot-->
Format Go code:
```suggestion
	"fmt"
	"os"

```

mattmoor commented on plugins/admin/core/root.go:
```diff
@@ -0,0 +1,99 @@
+	"knative.dev/client-contrib/plugins/admin/pkg/command"
+	"knative.dev/client-contrib/plugins/admin/pkg/command/domain"
+	private_registry "knative.dev/client-contrib/plugins/admin/pkg/command/private-registry"
+	"os"
```
<!--gotool-bot-->
Format Go code:
```suggestion

```

mattmoor commented on plugins/admin/pkg/command/domain/domain.go:
```diff
@@ -0,0 +1,37 @@
+package domain
+
+import (
+	"fmt"
```
<!--gotool-bot-->
Format Go code:
```suggestion
	"fmt"

```

mattmoor commented on plugins/admin/pkg/command/domain/set.go:
```diff
@@ -0,0 +1,106 @@
+
+import (
+	"fmt"
+	"knative.dev/client-contrib/plugins/admin/pkg"
```
<!--gotool-bot-->
Format Go code:
```suggestion

```

mattmoor commented on plugins/admin/pkg/command/domain/set.go:
```diff
@@ -0,0 +1,106 @@
+	"fmt"
+	"knative.dev/client-contrib/plugins/admin/pkg"
+	"os"
+	"strings"
```
<!--gotool-bot-->
Format Go code:
```suggestion
	"strings"

	"knative.dev/client-contrib/plugins/admin/pkg"
```

mattmoor commented on plugins/admin/pkg/command/private-registry/enable.go:
```diff
@@ -0,0 +1,127 @@
+package private_registry
+
+import (
+	"fmt"
```
<!--gotool-bot-->
Format Go code:
```suggestion
	"fmt"
	"os"

```

mattmoor commented on plugins/admin/pkg/command/private-registry/enable.go:
```diff
@@ -0,0 +1,127 @@
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"knative.dev/client-contrib/plugins/admin/pkg"
+	"os"
```
<!--gotool-bot-->
Format Go code:
```suggestion

```

mattmoor commented on plugins/admin/pkg/command/private-registry/enable.go:
```diff
@@ -0,0 +1,127 @@
+	"knative.dev/client-contrib/plugins/admin/pkg"
+	"os"
+
+	"encoding/json"
```
<!--gotool-bot-->
Format Go code:
```suggestion
	"encoding/json"

```

mattmoor commented on plugins/admin/vendor/github.com/hashicorp/hcl/decoder.go:
```diff
@@ -0,0 +1,729 @@
+	// we need to un-flatten the ast enough to decode
+	newNode := &ast.ObjectItem{
+		Keys: []*ast.ObjectKey{
+			&ast.ObjectKey{
```
<!--gotool-bot-->
Format Go code:
```suggestion
			{
```

mattmoor replied:
<!--gotool-bot-->
Format Go code:
```suggestion
			{
```

mattmoor commented on plugins/admin/vendor/github.com/modern-go/concurrent/log.go:
```diff
@@ -0,0 +1,13 @@
+package concurrent
+
+import (
+	"os"
+	"log"
```
<!--gotool-bot-->
Format Go code:
```suggestion

```

mattmoor replied:
<!--gotool-bot-->
Format Go code:
```suggestion

```

mattmoor commented on plugins/admin/vendor/github.com/modern-go/concurrent/log.go:
```diff
@@ -0,0 +1,13 @@
+import (
+	"os"
+	"log"
+	"io/ioutil"
```
<!--gotool-bot-->
Format Go code:
```suggestion
	"io/ioutil"
	"log"
	"os"
```

mattmoor replied:
<!--gotool-bot-->
Format Go code:
```suggestion
	"io/ioutil"
	"log"
	"os"
```

mattmoor commented on plugins/admin/vendor/github.com/modern-go/concurrent/log.go:
```diff
@@ -0,0 +1,13 @@
+// ErrorLogger is used to print out error, can be set to writer other than stderr
+var ErrorLogger = log.New(os.Stderr, "", 0)
+
+// InfoLogger is used to print informational message, default to off
```
<!--gotool-bot-->
Format Go code:
```suggestion
// InfoLogger is used to print informational message, default to off
var InfoLogger = log.New(ioutil.Discard, "", 0)
```

mattmoor replied:
<!--gotool-bot-->
Format Go code:
```suggestion
// InfoLogger is used to print informational message, default to off
var InfoLogger = log.New(ioutil.Discard, "", 0)
```

mattmoor commented on plugins/admin/vendor/github.com/modern-go/concurrent/unbounded_executor.go:
```diff
@@ -0,0 +1,119 @@
+
+import (
+	"context"
+	"fmt"
```
<!--gotool-bot-->
Format Go code:
```suggestion
	"fmt"
	"reflect"
```

mattmoor replied:
<!--gotool-bot-->
Format Go code:
```suggestion
	"fmt"
	"reflect"
```

mattmoor commented on plugins/admin/vendor/github.com/modern-go/concurrent/unbounded_executor.go:
```diff
@@ -0,0 +1,119 @@
+	"runtime/debug"
+	"sync"
+	"time"
+	"reflect"
```
<!--gotool-bot-->
Format Go code:
```suggestion

```

mattmoor replied:
<!--gotool-bot-->
Format Go code:
```suggestion

```

mattmoor commented on plugins/admin/vendor/github.com/modern-go/reflect2/reflect2.go:
```diff
@@ -0,0 +1,298 @@
+func (cfg Config) Froze() *frozenConfig {
+	return &frozenConfig{
+		useSafeImplementation: cfg.UseSafeImplementation,
+		cache: concurrent.NewMap(),
```
<!--gotool-bot-->
Format Go code:
```suggestion
		cache:                 concurrent.NewMap(),
```

mattmoor replied:
<!--gotool-bot-->
Format Go code:
```suggestion
		cache:                 concurrent.NewMap(),
```

mattmoor commented on plugins/admin/vendor/github.com/modern-go/reflect2/reflect2.go:
```diff
@@ -0,0 +1,298 @@
+	stringHeader := (*reflect.StringHeader)(unsafe.Pointer(&str))
+	sliceHeader := &reflect.SliceHeader{
+		Data: stringHeader.Data,
+		Cap: stringHeader.Len,
+		Len: stringHeader.Len,
```
<!--gotool-bot-->
Format Go code:
```suggestion
		Cap:  stringHeader.Len,
		Len:  stringHeader.Len,
```

mattmoor replied:
<!--gotool-bot-->
Format Go code:
```suggestion
		Cap:  stringHeader.Len,
		Len:  stringHeader.Len,
```

mattmoor commented on plugins/admin/vendor/github.com/spf13/afero/memmap.go:
```diff
@@ -0,0 +1,365 @@
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+
+	for p, _ := range m.getData() {
```
<!--gotool-bot-->
Format Go code:
```suggestion
	for p := range m.getData() {
```

mattmoor replied:
<!--gotool-bot-->
Format Go code:
```suggestion
	for p := range m.getData() {
```

mattmoor commented on plugins/admin/vendor/gopkg.in/yaml.v2/readerc.go:
```diff
@@ -0,0 +1,412 @@
+
+	// [Go] This function was changed to guarantee the requested length size at EOF.
+	// The fact we need to do this is pretty awful, but the description above implies
+	// for that to be the case, and there are tests 
```
<!--gotool-bot-->
Format Go code:
```suggestion
	// for that to be the case, and there are tests
```

mattmoor replied:
<!--gotool-bot-->
Format Go code:
```suggestion
	// for that to be the case, and there are tests
```

mattmoor commented on plugins/admin/vendor/gopkg.in/yaml.v2/resolve.go:
```diff
@@ -0,0 +1,258 @@
+					return yaml_INT_TAG, uintv
+				}
+			} else if strings.HasPrefix(plain, "-0b") {
+				intv, err := strconv.ParseInt("-" + plain[3:], 2, 64)
```
<!--gotool-bot-->
Format Go code:
```suggestion
				intv, err := strconv.ParseInt("-"+plain[3:], 2, 64)
```

mattmoor replied:
<!--gotool-bot-->
Format Go code:
```suggestion
				intv, err := strconv.ParseInt("-"+plain[3:], 2, 64)
```

mattmoor commented on plugins/admin/vendor/gopkg.in/yaml.v2/sorter.go:
```diff
@@ -0,0 +1,113 @@
+		var ai, bi int
+		var an, bn int64
+		if ar[i] == '0' || br[i] == '0' {
+			for j := i-1; j >= 0 && unicode.IsDigit(ar[j]); j-- {
```
<!--gotool-bot-->
Format Go code:
```suggestion
			for j := i - 1; j >= 0 && unicode.IsDigit(ar[j]); j-- {
```

mattmoor replied:
<!--gotool-bot-->
Format Go code:
```suggestion
			for j := i - 1; j >= 0 && unicode.IsDigit(ar[j]); j-- {
```

mattmoor commented on plugins/admin/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/meta.go:
```diff
@@ -0,0 +1,178 @@
+func (meta *ObjectMeta) SetDeletionTimestamp(deletionTimestamp *Time) {
+	meta.DeletionTimestamp = deletionTimestamp
+}
+func (meta *ObjectMeta) GetDeletionGracePeriodSeconds() *int64 { return meta.DeletionGracePeriodSeconds }
```
<!--gotool-bot-->
Format Go code:
```suggestion
func (meta *ObjectMeta) GetDeletionGracePeriodSeconds() *int64 {
	return meta.DeletionGracePeriodSeconds
}
```

mattmoor replied:
<!--gotool-bot-->
Format Go code:
```suggestion
func (meta *ObjectMeta) GetDeletionGracePeriodSeconds() *int64 {
	return meta.DeletionGracePeriodSeconds
}
```

mattmoor commented on plugins/admin/vendor/k8s.io/apimachinery/pkg/fields/selector.go:
```diff
@@ -0,0 +1,476 @@
+
+type nothingSelector struct{}
+
+func (n nothingSelector) Matches(_ Fields) bool                                      { return false }
+func (n nothingSelector) Empty() bool                                                { return false }
+func (n nothingSelector) String() string                                             { return "" }
+func (n nothingSelector) Requirements() Requirements                                 { return nil }
+func (n nothingSelector) DeepCopySelector() Selector                                 { return n }
+func (n nothingSelector) RequiresExactMatch(field string) (value string, found bool) { return "", false }
+func (n nothingSelector) Transform(fn TransformFunc) (Selector, error)               { return n, nil }
```
<!--gotool-bot-->
Format Go code:
```suggestion
func (n nothingSelector) Matches(_ Fields) bool      { return false }
func (n nothingSelector) Empty() bool                { return false }
func (n nothingSelector) String() string             { return "" }
func (n nothingSelector) Requirements() Requirements { return nil }
func (n nothingSelector) DeepCopySelector() Selector { return n }
func (n nothingSelector) RequiresExactMatch(field string) (value string, found bool) {
	return "", false
}
func (n nothingSelector) Transform(fn TransformFunc) (Selector, error) { return n, nil }
```

mattmoor replied:
<!--gotool-bot-->
Format Go code:
```suggestion
func (n nothingSelector) Matches(_ Fields) bool      { return false }
func (n nothingSelector) Empty() bool                { return false }
func (n nothingSelector) String() string             { return "" }
func (n nothingSelector) Requirements() Requirements { return nil }
func (n nothingSelector) DeepCopySelector() Selector { return n }
func (n nothingSelector) RequiresExactMatch(field string) (value string, found bool) {
	return "", false
}
func (n nothingSelector) Transform(fn TransformFunc) (Selector, error) { return n, nil }
```

mattmoor commented on plugins/admin/vendor/k8s.io/apimachinery/pkg/labels/selector.go:
```diff
@@ -0,0 +1,914 @@
+
+type nothingSelector struct{}
+
+func (n nothingSelector) Matches(_ Labels) bool                                      { return false }
+func (n nothingSelector) Empty() bool                                                { return false }
+func (n nothingSelector) String() string                                             { return "" }
+func (n nothingSelector) Add(_ ...Requirement) Selector                              { return n }
+func (n nothingSelector) Requirements() (Requirements, bool)                         { return nil, false }
+func (n nothingSelector) DeepCopySelector() Selector                                 { return n }
+func (n nothingSelector) RequiresExactMatch(label string) (value string, found bool) { return "", false }
```
<!--gotool-bot-->
Format Go code:
```suggestion
func (n nothingSelector) Matches(_ Labels) bool              { return false }
func (n nothingSelector) Empty() bool                        { return false }
func (n nothingSelector) String() string                     { return "" }
func (n nothingSelector) Add(_ ...Requirement) Selector      { return n }
func (n nothingSelector) Requirements() (Requirements, bool) { return nil, false }
func (n nothingSelector) DeepCopySelector() Selector         { return n }
func (n nothingSelector) RequiresExactMatch(label string) (value string, found bool) {
	return "", false
}
```

mattmoor replied:
<!--gotool-bot-->
Format Go code:
```suggestion
func (n nothingSelector) Matches(_ Labels) bool              { return false }
func (n nothingSelector) Empty() bool                        { return false }
func (n nothingSelector) String() string                     { return "" }
func (n nothingSelector) Add(_ ...Requirement) Selector      { return n }
func (n nothingSelector) Requirements() (Requirements, bool) { return nil, false }
func (n nothingSelector) DeepCopySelector() Selector         { return n }
func (n nothingSelector) RequiresExactMatch(label string) (value string, found bool) {
	return "", false
}
```

rhuss commented:
> The following users are mentioned in OWNERS file(s) but are not members of the knative org.
> 
> Once all users have been added as members of the org, you can trigger verification by writing `/verify-owners` in a comment.
> 
> * zhanggbj
>   
>   * OWNERS

@zhanggbj I asked the Knative admins to add you to the GitHub orga

zhanggbj commented:
Hi @rhuss @maximilien,
I've restructured kn-admin to align with the hello plugin, including folders/vendors/Reame/LICENSE/etc. would you please help to take a review? Thanks!

This is an initial PR for kn-admin and I still have a lot of TODOs in my list. I'll raise new PRs for review later as this already got a little bit hard to review:-)

- Fill up unit test and e2e test for kn-admin
- New features (detailes in the `docs` folder)

rhuss commented on OWNERS:
```diff
@@ -4,6 +4,7 @@ approvers:
 - rhuss
 - maximilien
 - navidshaikh
+- zhanggbj
```
I wonder whether we could add a OWNERS fils within the `plugins/admin` directory ? I think this would be nice as it highlights the responsibilities and point-of-contact better. 

I think prow supports this, let me check ...

zhanggbj replied:
Sounds good to me. Thanks a lot and I've joined Knative org.

rhuss commented on plugins/admin/README.adoc:
```diff
@@ -0,0 +1,126 @@
+  --docker-server=[PRIVATE_REGISTRY_SERVER_URL] \
+  --docker-email=[PRIVATE_REGISTRY_EMAIL] \
+  --docker-username=[PRIVATE_REGISTRY_USER] \
+  --docker-password=[PRIVATE_REGISTRY_PASSWORD]
```
I don't think it should be called `--docker` as this about a generic registry right (not only holding Docker images but OCI images in general). So I suggest to rename it to `--registry-url`, `--registry-email`, `--registry-user`, `--registry-password`

zhanggbj replied:
@rhuss I agree with you and removed `--docker`, and about the flags, I think its obvious here they're for registry, so I rename it like `--server` `--username` `--password` to make it shorter.

rhuss commented on plugins/admin/README.adoc:
```diff
@@ -0,0 +1,126 @@
+  admin private-registry [command]
+
+Available Commands:
+  enable      enable Service deployment from a private registry
```
I wonder whether we need another level of command here. (Also the options to this command are not shown here).
Also, I wouldn't call it "private-registry" (as it could be a public registry with password auth, too ?). So maybe "kn admin set-registry --...." or "kn admin registry set" (if you plan a "kn admin registry get", too).

rhuss replied:
Also you then could remove the "--docker" in the options as it is evident for what it is.

zhanggbj replied:
@rhus
I agree with you. I removed `private` here.
About the command name, I think instead of `enable` or `set`, maybe `add` is more proper here, as user may add not only one registry for its platform, and later I'll add `registry list` and `registry rm` etc.
And `--docker ` is removed as above comment

rhuss commented on plugins/admin/README.adoc:
```diff
@@ -0,0 +1,126 @@
+.Update the default route domain if --selector no specified
+====
+----
+$ Kn admin domain set --custom-domain mydomain.com
```
```suggestion
$ kn admin domain set --custom-domain mydomain.com
```

zhanggbj replied:
fixed

rhuss commented on plugins/admin/README.adoc:
```diff
@@ -0,0 +1,126 @@
+.Update a custom domain with --selector and Service with a label app=v1 will use test.com
+====
+----
+$ Kn admin domain set --custom-domain test.com --selector app=v1
```
```suggestion
$ kn admin domain set --custom-domain test.com --selector app=v1
```

zhanggbj replied:
fixed

rhuss commented on plugins/admin/core/root.go:
```diff
@@ -0,0 +1,99 @@
+
+For example:
+kn admin domain set - to set Knative route domain
+kn admin private-registry enable - to enable deployment from the private registry
```
Not sure why it "set" and "enable" for both being write operation. With "enable" I would expect also a "disable", but when we do have "set" then we could also set a "null" URL to indicate to remove a registry. Also, as mentioned I just would call it "registry".

zhanggbj replied:
@rhuss 
As in above comment, I updated to `kn admin registry add`, I think this is more proper here, and later we can discuss more to add `registry list` or `registry rm`.

rhuss commented:
@zhanggbj btw, I would be happy to already commit the PR and then we can continue to work on it later.

zhanggbj commented on plugins/admin/README.adoc:
```diff
@@ -0,0 +1,126 @@
+### Usage
+
+----
+A plugin of kn client to manage Knative for administrators.
```
fixed!

zhanggbj commented on plugins/admin/README.adoc:
```diff
@@ -0,0 +1,126 @@
+kn admin domain set - to set Knative route domain
+
+Usage:
+  admin domain [flags]
+  admin domain [command]
```
fixed

zhanggbj commented on plugins/admin/README.adoc:
```diff
@@ -0,0 +1,126 @@
+  --docker-password=[PRIVATE_REGISTRY_PASSWORD]
+
+Usage:
+  admin private-registry [command]
```
fixed

zhanggbj commented on plugins/admin/README.adoc:
```diff
@@ -0,0 +1,126 @@
+kn admin domain set - to set Knative route domain
+kn admin private-registry enable - to enable deployment from the private registry
+
+Usage:
```
Fixed!

zhanggbj commented on plugins/admin/README.adoc:
```diff
@@ -0,0 +1,126 @@
+
+A kn plugin for Knative cluster management.
+
+`kn hello` prints out a friendly message to the user.
```
fixed!

zhanggbj commented:
@rhuss 
I've addressed all your comments and would you please help to take a look? Thanks!
And this is an initial PR, i think there're still a lot of todos for new subcommand and for existed `domain` and `registry` CRUD actions, and we can discuss later after this PR is merged and easy for collaboration.
